pipeline/data/importers/mono/hplt.py

```diff
@@ -105,129 +133,170 @@ def load_shuffled_shard_urls(language: str) -> list[str]:
+        self.visited_lines = 0
+        self.file_destination = file_destination
+        self.stats = FilteringStatistics(file_destination)
+        self.strings_seen = WeakStringSet()
+        self.stack = ExitStack()
+        self.outfile = self.stack.enter_context(write_lines(file_destination))
+
+    def close(self):
+        self.stack.close()
+
+    def download(self):
+        try:
+            self._run_download()
+        finally:
+            self.close()
+
+    def _run_download(self):
+        logger.info(f"Using HPLT locale {self.hplt_locale}")
+        shuffled_shard_urls = load_shuffled_shard_urls(self.hplt_locale)
+        self.stats.shards.filtered = len(shuffled_shard_urls)
 
         # The shard URLs are shuffled, and then streamed into the read_lines iterator.
         # This iterator can work over multiple documents. The first document is loaded,
         # and then the documents in the shard are read in order from that shard. After
         # the first shard is read, the iterator continues with the next shards until
         # enough fluent sentences are collected. At this point the remaining shards
         # will not be visited.
-        document_stream = stack.enter_context(
-            read_lines(shuffled_shard_urls, on_enter_location=stats.count_shards_visited)
+        document_stream = self.stack.enter_context(
+            read_lines(shuffled_shard_urls, on_enter_location=self.stats.count_shards_visited)
         )
 
-        strings_seen = WeakStringSet()
-        accumulated_text: str = ""
-        cumulative_char_count = 0
-        visited_lines = 0
-
-        def maybe_write_accumulated_text():
-            """
-            Since the loop below is building up paragraphs of text, we only want to write
-            out a line when enough text has been accumulated. The paragraph should be
-            written out when either the text gets too long, or the next line is discarded.
-            """
-            nonlocal accumulated_text
-            nonlocal cumulative_char_count
-            cumulative_char_count = 0
-            if accumulated_text:
-                if accumulated_text in strings_seen:
-                    stats.duplicate_lines.value += 1
-                else:
-                    outfile.write(accumulated_text + "\n")
-                    stats.final_lines.value += 1
-                    strings_seen.add(accumulated_text)
-                accumulated_text = ""
-
         for document_json in document_stream:
-            stats.document_count.value += 1
-
+            self.stats.document_count.value += 1
             document = HPLTDocument(**json.loads(document_json))
+            overall_doc_score = document.doc_scores[0]
+            doc_lang = document.lang[0]
+
+            self._maybe_write_accumulated_text()
+
+            # HPLT 2.0 uses document level scores
+            if overall_doc_score < self.hplt_min_doc_score:
+                self.stats.filtered_doc_score.value += 1
```

Comment:
Thanks for adding the statistics counting.

Reply:
The detailed stats json is very helpful to understand what's happening with filtering. Thank you for adding support for this! I think we should output such json files for all filtering steps we have but it will require refactoring bash to python...